src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -405,13 +481,98 @@ static private void FlattenDeconstructVariables(ImmutableArray<DeconstructionVar
+
+            ArrayBuilder<DeconstructionVariable> variables = BindDeconstructionDeclarationVariables(node.Declaration, null, diagnostics);
+
+            var result = new BoundExpressionStatement(node, BindDeconstructionAssignment(node.Declaration.Deconstruction.Value, node.Declaration.Deconstruction.Value, variables, diagnostics));
+            FreeDeconstructionVariables(variables);
+
+            return result;
+        }
+
+        /// <summary>
+        /// The caller is responsible for releasing the nested ArrayBuilders.
+        /// </summary>
+        private ArrayBuilder<DeconstructionVariable> BindDeconstructionDeclarationVariables(VariableDeclarationSyntax node, TypeSyntax varSyntax, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction != null);
+            var variables = node.Deconstruction.Variables;
+
+            if (varSyntax == null)
+            {
+                varSyntax = node.Type;
+            }
+
+            // There are four cases for VariableDeclaration:
+            // - type and declarators are set, but deconstruction is null. This could represent `int x`, which is a single variable.
+            // - type is null, declarators are set, but deconstruction is null. This could represent `x`, which is a single variable.
+            // - type is set to 'var', declarators are null, and deconstruction is set. This could represent `var (...)`
+            // - type and declarators are null, but deconstruction is set. Thi could represent `(int x, ...)`
+
+            var localsBuilder = ArrayBuilder<BoundLocalDeclaration>.GetInstance(variables.Count);
+            var variablesBuilder = ArrayBuilder<DeconstructionVariable>.GetInstance(variables.Count);
+            foreach (var variable in variables)
+            {
+                if (variable.Deconstruction == null)
+                {
+                    variablesBuilder.Add(new DeconstructionVariable(BindDeconstructionDeclarationVariable(variable, varSyntax, localsBuilder, diagnostics)));
+                }
+                else
+                {
+                    variablesBuilder.Add(new DeconstructionVariable(BindDeconstructionDeclarationVariables(variable, varSyntax, diagnostics)));
+                }
+            }
+
+            return variablesBuilder;
+        }
+
+        /// <summary>
+        /// Returns a BoundLocal when the type was explicit, otherwise returns a DeconstructionLocalPendingInference.
+        /// </summary>
+        private BoundExpression BindDeconstructionDeclarationVariable(VariableDeclarationSyntax node, TypeSyntax varSyntax, ArrayBuilder<BoundLocalDeclaration> locals, DiagnosticBag diagnostics)
+        {
+            Debug.Assert(node.Deconstruction == null);
+            Debug.Assert(varSyntax != null || node.Type != null);
+            Debug.Assert(node.Variables.Count == 1);
+
+            var declarator = node.Variables[0];
+            var typeSyntax = varSyntax ?? node.Type;
```

Comment:
I know it is an error, but still using the closes non-null type syntax, even in error recovery case is, better I think. In the example that you gave, type of y should be int and type of z should be string, we shouldn't try to infer them. And don't forget that 'var' isn't var until proven so. So, 'var (x, (y, z))' could be invalid too and the type could actually be explicit in this case. BTW, do we have a unit-test for a scenario like that?

---
In reply to: [68813077](https://github.com/dotnet/roslyn/pull/12224#discussion_r68813077) [](ancestors = 68813077)